### PR TITLE
feat: Implement Spark-compatible CAST from String to Date

### LINKED ISSUE #327: Implement Spark-compatible CAST from String to Date
### What is the problem the feature request solves?

# What is the problem the feature request solves?

We currently delegate to DataFusion when casting from string to date and there are some differences in behavior compared to Spark.

- Spark supports dates in the format `YYYY` and `YYYY-MM` and DataFusion does not
- Spark supports a trailing `T` as in `2024-01-01T` and DataFusion does not
- DataFusion doesn't throw an exception for invalid inputs in ANSI mode

Edge cases from fuzz testing: 

| Input | Spark | DataFusion |
|-|-|-|
| `"-0973250"` | `3251-01-01` | `null` |
| `"-3638-5"` | `3639-05-01` | `null` | 

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by vidyasankarv**
## Which issue does this PR close?

Closes #327.

### CODE REVIEW

File: core/src/execution/datafusion/expressions/cast.rs
Comment (parthchandra) [inline core/src/execution/datafusion/expressions/cast.rs:111]: Any reason why this is a macro and not a function? I see only one usage (maybe I missed something).

File: core/src/execution/datafusion/expressions/cast.rs
Comment (parthchandra) [inline core/src/execution/datafusion/expressions/cast.rs:304]: Seems like unnecessary re-formatting (multiple places).

File: core/src/execution/datafusion/expressions/cast.rs
Comment (vidyasankarv) [reply core/src/execution/datafusion/expressions/cast.rs]: Hi @parthchandra thanks for the review. I am new to rust (and also my first attempt at an oss contribution). 
i was imitating the code from cast_utf8_to_timestamp. removed it now.

File: core/src/execution/datafusion/expressions/cast.rs
Comment (vidyasankarv) [reply core/src/execution/datafusion/expressions/cast.rs]: cleaned this up now with cargo fmt

**Comment by vidyasankarv** [comment]
This PR is still in progress. I added support for String to Date32. Do we have to support Date64 as well.

* Spark supports dates in the format YYYY and YYYY-MM and DataFusion does not - supported now
* Spark supports a trailing T as in 2024-01-01T and DataFusion does not - supported now
* DataFusion doesn't throw an exception for invalid inputs in ANSI mode - returns error in ANSI mode if date cant be parsed.

Hi  @parthchandra @andygrove  Can you please review if this is going in the right direction.

Questions:
* Need some pointers on fuzz tests. If I understand correctly 
* "-0973250", "-3638-5" fuzz tests in Legacy mode should return values as mentioned https://github.com/apache/datafusion-comet/issues/327 - currently legacy mode returns null.

**Comment by andygrove** [comment]
> This PR is still in progress. I added support for String to Date32.
> 
> * Spark supports dates in the format YYYY and YYYY-MM and DataFusion does not - supported now
> * Spark supports a trailing T as in 2024-01-01T and DataFusion does not - supported now
> * DataFusion doesn't throw an exception for invalid inputs in ANSI mode - returns error in ANSI mode if date cant be parsed.
> 
> Hi @parthchandra @andygrove Can you please review if this is going in the right direction.
> 
> Questions:
> 
> * Need some pointers on fuzz tests. If I understand correctly
>   
>   * "-0973250", "-3638-5" fuzz tests in Legacy mode should return values as mentioned [Implement Spark-compatible CAST from String to Date #327](https://github.com/apache/datafusion-comet/issues/327) - currently legacy mode returns null.
> * Do we have to support Date64 as well ?

vidyasankarv Yes, I would say this is going it a good direction based on a very quick review. I will try and find more time tomorrow for a deeper look. To answer your questions:

- Yes, ideally we should support the edge cases mentioned in the issue. We could also choose to leave that for a future PR and leave the current support marked as incompatible and provide some documentation on what is not supported (as we have done for string -> timestamp).
- No need to support Date64 for this issue

File: core/src/execution/datafusion/expressions/cast.rs
Comment (parthchandra) [inline core/src/execution/datafusion/expressions/cast.rs:996]: I wasn't familiar with Spark's string to date conversion so I took a look. (https://github.com/apache/spark/blob/9d79ab42b127d1a12164cec260bfbd69f6da8b74/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L312)
From the comment the allowed formats are - 
```
  * `[+-]yyyy*`
   * `[+-]yyyy*-[m]m`
   * `[+-]yyyy*-[m]m-[d]d`
   * `[+-]yyyy*-[m]m-[d]d `
   * `[+-]yyyy*-[m]m-[d]d *`
   * `[+-]yyyy*-[m]m-[d]dT*`
  ```  
  I honestly don't know what a string with a 'plus/minus' at the beginning of the date even means but you might want to handle that case.
  Also, the max number of digits allowed for the year is 7. 
  Finally, once you've got the 'day' segment of the date you may have a ' ' or 'T' (you're only handling the latter) and the characters after that are discarded.
   It looks to me like Spark's custom implementation might be slightly faster since it manages to achieve the split of the string into segments and the parsing of the digits in a single pass. (also does not need to prepare the parser with the format string). You might want to consider doing the same.

**Comment by parthchandra** [comment]
> * "-0973250", "-3638-5" fuzz tests in Legacy mode should return values as mentioned [Implement Spark-compatible CAST from String to Date #327](https://github.com/apache/datafusion-comet/issues/327) - currently legacy mode returns null.

You're pretty close to handling these cases (see my comment in the review)

File: core/src/execution/datafusion/expressions/cast.rs
Comment (parthchandra) [reply core/src/execution/datafusion/expressions/cast.rs]: > I am new to rust

So am I :). Welcome to the community !

**Comment by vidyasankarv** [comment]
hi @parthchandra @andygrove made changes as suggested
ported the date parsing logic [from SparkDateTimeUtils](https://github.com/apache/spark/blob/9d79ab42b127d1a12164cec260bfbd69f6da8b74/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L312)

This PR does not support fuzzy tests in CometCastSuite for `test Date to String` as Naive Date only supports dates in the below range and the dates generated by fuzzy test for matching with spark are out of this range. 
```
/// The minimum possible `Na iveDate` (January 1, 262145 BCE).
#[deprecated(since = "0.4.20", note = "Use NaiveDate::MIN instead")]
pub const MIN_DATE: NaiveDate = NaiveDate::MIN;
/// The maximum possible `NaiveDate` (December 31, 262143 CE).
#[deprecated(since = "0.4.20", note = "Use NaiveDate::MAX instead")]
pub const MAX_DATE: NaiveDate = NaiveDate::MAX;

```

For Ansi for any format validatoin errors
* Try/Legacy mode return - None
* Ansi - Err

However when all validations pass, for dates that are beyond dates supported by Native Mode all modes return None.

Can you please take another look at the PR. thank you.

**Comment by andygrove** [comment]
Thanks @vidyasankarv. I plan on carefully reviewing this later today.

File: core/src/execution/datafusion/expressions/cast.rs
Comment (andygrove) [inline core/src/execution/datafusion/expressions/cast.rs:1572]: This line triggers an overflow panic if `b` is less than `0`. It looks like this code assumes that `b` is a digit, but with the input `3/`, it failed here on processing `/`. Perhaps check that `b` is a digit first?

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:570]: Could you add some invalid entries here as well so that we can ensure that Comet throws errors for invalid inputs when ANSI is enabled?

Some suggestions:

```scala
        "",
        "0",
        "not_a_date",
```

File: spark/src/main/scala/org/apache/comet/expressions/CometCast.scala
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/expressions/CometCast.scala:122]: Although it seems we are compatible for most common use cases, we are not 100% compatible, so should add a note here.

```suggestion
        Compatible(Some("Only supports years between 262143 BC and 262142 AD"))
```

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (kazuyukitanimura) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:568]: Should we keep `generateStrings` because that covers random value tests?

File: spark/src/main/scala/org/apache/comet/expressions/CometCast.scala
Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/comet/expressions/CometCast.scala:122]: If we do not cover these cases in this PR, we should add a test with `ignore`

Comment (vidyasankarv) [reply spark/src/main/scala/org/apache/comet/expressions/CometCast.scala]: changed Compatible as suggested
also added back an ignore test case for fuzz test as a placeholder

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (vidyasankarv) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala]: added additional negative test cases as suggested.

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (vidyasankarv) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:568]: `generateStrings` method wasnt removed, just the fuzz test for Dates we removed, but now added back the fuzz test for Dates as an ignore test  some dates are not supported.

File: core/src/execution/datafusion/expressions/cast.rs
Comment (vidyasankarv) [reply core/src/execution/datafusion/expressions/cast.rs]: @andygrove added check for ascii digits and some negative test cases around this in rust code. thank you

**Comment by vidyasankarv** [comment]
Hi @andygrove 
I need some help 
https://github.com/apache/datafusion-comet/pull/383/files#diff-41ecdd113d7a7afe33447e34f1ff0b5ed3033a89bfbcefa9e7e259d7a6e4daecR593

These particular test cases result in failure - like when i run with this sample test `2020-10-010T` - the same test in rust returns `CometError` in ansi mode https://github.com/apache/datafusion-comet/pull/383/files#diff-b7339cca414a6315488506dd33654946f62c229feb8ad0d4abeda683ca75b4b5R1717

However when the test suite runs it fails for the combination of comet ansi enabled without try_cast

I have tried to concurrently debug on CLION https://github.com/apache/datafusion-comet/blob/main/docs/source/contributor-guide/debugging.md - however the breakpoints show us disabled and are nt hitting - i tried switching to lldb in the clion toolchain too but no help.

I added some additional logging locally and can see that it returns a CometError for the invalid value on the rust side.

Is there something else I could be missing in terms of any configuration. Appreciate any help when you get some time.

Thank you

File: core/src/execution/datafusion/expressions/cast.rs
Comment (andygrove) [inline core/src/execution/datafusion/expressions/cast.rs:663]: @vidyasankarv This is the fix you need to make your current test pass.

The problem was that we were ignoring any error here when runnin in ANSI mode.

```suggestion
                    } else if let Some(cast_value) = date_parser(string_array.value(i), eval_mode)?
```

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:610]: Let's add fuzzing back here, but filter out values that we know that we cannot support.

```suggestion
    // due to limitations of NaiveDate we only support years between 262143 BC and 262142 AD"
    // we can't test all possible fuzz dates
    val unsupportedYearPattern: Regex = "^\\s*[0-9]{5,}".r
    val fuzzDates = generateStrings(datePattern, 8)
      .filterNot(str => unsupportedYearPattern.findFirstMatchIn(str).isDefined)
    castTest((validDates ++ invalidDates ++ fuzzDates).toDF("a"), DataTypes.DateType)
```

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:568]: I left a suggestion for adding the fuzzing back but filtering out values that we know are not supported

File: core/src/execution/datafusion/expressions/cast.rs
Comment (kazuyukitanimura) [inline core/src/execution/datafusion/expressions/cast.rs:1489]: Is this defined because we cannot use `String#trim()` or `trim_matches`?

File: core/src/execution/datafusion/expressions/cast.rs
Comment (vidyasankarv) [reply core/src/execution/datafusion/expressions/cast.rs:663]: oh sorry my mistake missed checking here.
handling all cases of return value of date_parser using matching clause now. 
thank you @andygrove

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (vidyasankarv) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:610]: thank you for the suggestion @andygrove  incorporated this as suggested.

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (vidyasankarv) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:569]: the ignore test has been done away with as per @andygrove 's suggestion above using a filtered fuzz test, so removed this now.

File: core/src/execution/datafusion/expressions/cast.rs
Comment (vidyasankarv) [reply core/src/execution/datafusion/expressions/cast.rs:1489]: leave as is for now based on above comment from @andygrove

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala
Comment (vidyasankarv) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:568]: removed now and added a filtered fuzzy test

**Comment by andygrove** [comment]
There is one test failure with JDK 8 / Spark 3.2:

```
- cast StringType to DateType *** FAILED *** (349 milliseconds)
  "[CAST_INVALID_INPUT] The value '0' of the type "STRING" cannot be cast to "DATE" because it is malformed. Correct the value as per the syntax, or change its target type. Use `try_cast` to tolerate malformed input and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error." did not contain "annot cast 0 to DateType." (CometCastSuite.scala:1069)
  ```

**Comment by vidyasankarv** [comment]
> There is one test failure with JDK 8 / Spark 3.2:
> 
> ```
> - cast StringType to DateType *** FAILED *** (349 milliseconds)
>   "[CAST_INVALID_INPUT] The value '0' of the type "STRING" cannot be cast to "DATE" because it is malformed. Correct the value as per the syntax, or change its target type. Use `try_cast` to tolerate malformed input and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error." did not contain "annot cast 0 to DateType." (CometCastSuite.scala:1069)
> ```

Hi @andygrove is there any example as to how we fix an issue that only happens for a specific version. thank you

**Comment by andygrove** [comment]
@vidyasankarv I would suggest that we skip the test for now when running against Spark 3.2 and file a follow on issue to fix 3.2 compatibility (this may not be a high priority since 3.2 is quite old and we should consider dropping support for it at some point.

You can add an `assume` call to the test to skip for certain criteria:

```scala
  test("cast StringType to DateType") {
    assume(CometSparkSessionExtensions.isSpark33Plus)
```

It would be good to add a comment in here as well with a link to the follow on issue (could you file that?)

**Comment by vidyasankarv** [comment]
> @vidyasankarv I would suggest that we skip the test for now when running against Spark 3.2 and file a follow on issue to fix 3.2 compatibility (this may not be a high priority since 3.2 is quite old and we should consider dropping support for it at some point.
> 
> You can add an `assume` call to the test to skip for certain criteria:
> 
> ```scala
>   test("cast StringType to DateType") {
>     assume(CometSparkSessionExtensions.isSpark33Plus)
> ```
> 
> It would be good to add a comment in here as well with a link to the follow on issue (could you file that?)

@andygrove thank you for suggestions - filed this issue https://github.com/apache/datafusion-comet/issues/440 and linked in the test.

**Comment by vidyasankarv** [comment]
https://github.com/apache/datafusion-comet/suites/23883332179/logs?attempt=2

In the logs for ubuntu-latest/java 17-spark-3.4-scala-2.12/java - which included the fuzz test dates - all return cast values on comet side are showing null.

Additionally sample dates that pass otherwise without fuzz test https://github.com/apache/datafusion-comet/pull/383/files#diff-41ecdd113d7a7afe33447e34f1ff0b5ed3033a89bfbcefa9e7e259d7a6e4daecR577-R585  also report as null on Comet side when the fuzz tests are included.

https://github.com/apache/datafusion-comet/actions/runs/9123082449/job/25132235801

```
  ![262142-01-01,2142-01-01]               [262142-01-01,null]
  ![262142-01-01 ,2142-01-01]              [262142-01-01 ,null]
  ![262142-01-01T ,2142-01-01]             [262142-01-01T ,null]
  ![262142-01-01T 123123123,2142-01-01]    [262142-01-01T 123123123,null]
   [263,null]                              [263,null]
```
if you search for `262142-01-01` in the logs you can see it reports as failing on lines 14035 to 14038 as above

similarly if you also search  for dates `-262143-12-31` on lines 10167 to 10171 
```
   [--262143-12-31,null]                   [--262143-12-31,null]
   [--262143-12-31T 1234 ,null]            [--262143-12-31T 1234 ,null]
  ![-262143-12-31,2144-12-31]              [-262143-12-31,null]
  ![-262143-12-31 ,2144-12-31]             [-262143-12-31 ,null]
  ![-262143-12-31T,2144-12-31]             [-262143-12-31T,null]
  ![-262143-12-31T ,2144-12-31]            [-262143-12-31T ,null]
  ![-262143-12-31T 123123123,2144-12-31]   [-262143-12-31T 123123123,null]
```
the return value of comet is reported as null

I have spent a fair time trying to understand why this is happening 
and also added some more samples from fuzzy dates into my current unit tests in rust tests and CometCastSuite
all of them are passing.

So I might need some help in trying to identify the issue with fuzz test.
Apologies for taking your time on this again.

**Comment by andygrove** [comment]
> So I have pushed this build removing the fuzz test now to see if the build passes. So I might need some help in trying to identify the issue with fuzz test.
> Apologies for taking your time on this again. Thank you for your help

Thanks @vidyasankarv. I plan on looking into this tomorrow. Overall, the PR looks good.

**Comment by andygrove** [comment]
@vidyasankarv I am also very confused .. values that fail in the fuzz test work in the other test 🤔 

I am debugging and will let you know when I get to the bottom of this mystery

**Comment by vidyasankarv** [comment]
@andygrove thank you very much for looking into this. tested fuzz test with your suggestions and is working now. pushed changes in the latest commit https://github.com/apache/datafusion-comet/pull/383/commits/88af45c7757854cfc94e628e8338f299c2aec5d7.

**Review by andygrove — APPROVED** [review]
LGTM pending CI. Thank you for your patience @vidyasankarv.

Once this is merged I will rebase https://github.com/apache/datafusion-comet/pull/461 which would have prevented some of the issues we ran into on this PR

**Comment by vidyasankarv** [comment]
@andygrove @parthchandra @kazuyukitanimura thank you for reviews and support in helping me through my first open source contribution. Its been a great learning experience. Still trying to grasp all the new things I learnt from this seemingly simple `good first issue`. My first exposure to rust, seeing JNI in action for interactions between spark and comet using Arrow. Hope to keep contributing. And waiting for @andygrove's updated version of his `How Query Engine's Work' . Thank you all.

**Pull request merged**
